### ISSUE TITLE
`test_lstm` in `quantization.bc.test_backward_compatibility.TestSerialization` fails on Intel Cascade Lake machines

### ISSUE THREAD

**Issue by imaginary-person** [pytorch/pytorch#59098]
## TLDR
This issue has been established to be hardware-dependent, as it can be reproduced with the current master branch's source-code as well. Please refer to #59100. Debugging this issue can potentially uncover some latent issues with the Intel Cascade Lake series (gcc 9.3 or the hypervisor, etc not supporting it well, as per Occam's razor?).

## 🐛 Bug

Hello, in #56992, I'm trying to add AVX512 support to ATen.
`test_lstm` in `quantization.bc.test_backward_compatibility.TestSerialization` is failing on CI (please read Additional Context) for AVX512 kernels in the CI check [`pytorch_linux_bionic_py3_8_gcc9_coverage_test1`](https://app.circleci.com/pipelines/github/pytorch/pytorch/328437/workflows/a7e8eb6b-0a4b-4fe9-acb7-c389307e29a9/jobs/13767215).

However, if I clone the repo locally and build, I don't encounter this failure on machines with AVX512 support.
Even if I install the [CI build artifact](https://13766736-65600975-gh.circle-artifacts.com/0/home/circleci/project/dist/torch-1.9.0a0%2Bgitd895ec3-cp38-cp38-linux_x86_64.whl) locally, I still don't encounter this failure. Please help me figure out the source of this error on CI. Thanks!

## Reproduction steps (require a machine with AVX512 support)

1. Either install the [CI build artifact](https://app.circleci.com/pipelines/github/pytorch/pytorch/328437/workflows/a7e8eb6b-0a4b-4fe9-acb7-c389307e29a9/jobs/13766736/artifacts) locally with,
`wget https://13766736-65600975-gh.circle-artifacts.com/0/home/circleci/project/dist/torch-1.9.0a0%2Bgitd895ec3-cp38-cp38-linux_x86_64.whl`
`pip install torch-1.9.0a0+gitd895ec3-cp38-cp38-linux_x86_64.whl`.

OR

Clone the repo of the PR locally,
  `git clone --recursive https://github.com/imaginary-person/pytorch-1 -b only_vec .`
  `USE_MKL=1 USE_MKLDNN=1 USE_LLVM=/usr/lib/llvm-10 python setup.py develop`

2. Run the test with,
`python test/test_quantization.py -v -k test_lstm`
Two tests will run with this keyword.
The test [failing on CI](https://app.circleci.com/pipelines/github/pytorch/pytorch/328437/workflows/a7e8eb6b-0a4b-4fe9-acb7-c389307e29a9/jobs/13767215/tests#failed-test-1) is `test_lstm` of `quantization.bc.test_backward_compatibility.TestSerialization`.

The test fails with error,
```
File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_quantized.py", line 161, in test_fn
    qfunction(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/quantization/bc/test_backward_compatibility.py", line 230, in test_lstm
    self._test_op(mod, input_size=[4, 4, 3], input_quantized=False, generate=False, new_zipfile_serialization=True)
  File "/var/lib/jenkins/workspace/test/quantization/bc/test_backward_compatibility.py", line 76, in _test_op
    self.assertEqual(qmodule(input_tensor), expected, atol=prec)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1388, in assertEqual
    self.assertEqual(x_, y_, atol=atol, rtol=rtol, msg=msg,
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1355, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 13 element(s) (out of 112) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.9640435565029293 (4.41188467448228e-06 vs. 0.9640479683876038), which occurred at index (3, 0, 6).
```

## Expected behavior

The test should pass on a machine with AVX512 support, as it does locally.

## Failing Environment

This issue has been established to be hardware-dependent, as it can be reproduced with the current master branch's source-code as well, which doesn't yet have AVX512-vectorized ATen kernels. Please refer to #59100.

## Environments in which the test passes (<strike>locally</strike> on a cloud platform with full-metal hardware)

I tried to reproduce the issue on two different machines with Ubuntu 18.04 & Ubuntu 20.04, respectively.
Before testing, I uninstalled previous PyTorch versions.
While installing the [CI build artifact](https://app.circleci.com/pipelines/github/pytorch/pytorch/328437/workflows/a7e8eb6b-0a4b-4fe9-acb7-c389307e29a9/jobs/13766736/artifacts), the installation took a few seconds (which is normal, as it suggests that the build didn't have to take place again).

After installing the CI [`pytorch_linux_bionic_py3_8_gcc9_coverage_build`](https://app.circleci.com/pipelines/github/pytorch/pytorch/328437/workflows/a7e8eb6b-0a4b-4fe9-acb7-c389307e29a9/jobs/13766736/artifacts) artifact,

1. Ubuntu 18.04 machine (with bare-metal Intel Xeon Silver 4114 sockets, Skylake)
<details>
PyTorch version: 1.9.0a0+gitd895ec3

OS: Ubuntu 18.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-11ubuntu0~18.04.1) 9.3.0
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8 (64-bit runtime)
Python platform: Linux-4.15.0-137-generic-x86_64-with-glibc2.27

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.9.0a0+gitd895ec3
</details>

2. Ubuntu 20.04 machine (with bare-metal Intel Xeon Gold 6142 sockets, Skylake)
<details>
OS: Ubuntu 20.04 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.4.0-67-generic-x86_64-with-glibc2.29
</details>

## Additional context

Currently, Linux CI checks that are tests are run with `large` `resource_class` VMs, which, in turn, are run on machines without AVX512 support. For `pytorch_linux_bionic_py3_8_gcc9_coverage_test1` & `pytorch_linux_bionic_py3_8_gcc9_coverage_test2`, I temporarily modified CircleCI config in that PR to use `xlarge` VMs, as they are run on machines with AVX512 support, so AVX512 kernels can be tested. This can be verified from the fact that `ATEN_CPU_CAPABILITY` for these modified CI checks was `avx512` in the CI logs.

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

**Comment by imaginary-person** [comment]
@supriyar @jerryzh168,

When #43209 was reported, this issue was irreproducible at your end with AVX2 vectorization.
as it was only reproducible on some specific hardware (reportedly, Intel Cascade Lake machines).
All Linux CI checks that are tests are run on `large` `resource_class` VMs, which in turn, are run on old machines.

In #56992, I used `xlarge` `resource_class` VMs for testing AVX512 kernels in 2 CI checks.
I'm guessing these VMs are run on Intel Cascade Lake machines, on which this issue was occurring for AVX2 kernels when #43209 was reported.

If you'd like to debug this issue, you can trigger CI checks in #56992 & SSH to the CI machine running `pytorch_linux_bionic_py3_8_gcc9_coverage_test1`.

Maybe Intel folks, such as @mingfeima, can also help out.

**Comment by imaginary-person** [comment]
Hello @supriyar, @jerryzh168 & @mingfeima, [on the current master branch's source-code](https://app.circleci.com/pipelines/github/pytorch/pytorch/328568/workflows/5d08cb07-c1b1-4f14-bac4-1975d972712b/jobs/13775111), this issue can be reproduced with an `xlarge` executor in #59100.

 `test_lstm (quantization.bc.test_backward_compatibility.TestSerialization) ... FAIL (0.046s)`

Hence it's been established that this issue is hardware-dependent.
While CircleCI uses AWS, even GCP & Microsoft Azure have procured a lot of Intel Cascade Lake machines in the last 2 years, and usually offer them as distinct instance-classes. So, debugging this issue is probably worth it.

cc @ezyang @malfet

**Comment by imaginary-person** [comment]
@jerryzh168 @supriyar @mingfeima @CaoZhongZ

The machine is [definitely a Cascade Lake](https://app.circleci.com/pipelines/github/pytorch/pytorch/328734/workflows/8ce2d195-f481-477d-b4ca-56b1352152f1/jobs/13783028) one as it has `avx512_vnni` in the `lscpu` output, which was introduced in Cascade Lake.
Ice Lake also supports this instruction set, but Ice Lake Xeon SP machines were only launched last month.
The hypervisor is KVM.

**Comment by imaginary-person** [comment]
It's possible that this issue might not reproduce with a more recent gcc version. GCC 9.3.0 was used to create the build on which this test fails on Cascade Lake machines, and that GCC version was released in March 2020 (so it's not quite old).

However, if this test would fail even with gcc 11.1 on Cascade Lake machines, then it'd probably be a gcc bug.

**Comment by imaginary-person** [comment]
Hello @hjl-tools, is it possible for you to request your compiler team (that submits patches to gcc) to look into this issue that's affecting gcc versions up to 9.3 on Intel Cascade Lake machines? If this issue persists even on gcc 11.1, then it should be patched.

**Comment by ezyang** [comment]
Hey @imaginary-person, two Qs for you.

First, what do you need on your end to unblock? Skipping the test as known broken on Cascade seems sufficient to shut up CI and let you continue to test on xlarge.

Second, any chance we can isolate this to a simple C program or assembly that gets misrun on Cascade Lake? I'd go about this by minimizing the test case, first tracing it down to a single operator from the test, and then spelunking into the operator implementation to get out the specific code that is responsible for doing the computation that is turning out incorreclty.

**Comment by imaginary-person** [comment]
Hello @ezyang,

1. Thanks for confirming that it'd be okay for me to skip that test case.
BTW, I was wondering if some silent-correction errors could also be happening with PyTorch built with gcc 9.3 on these machines,
but that's a digression from implementing AVX512.

2. Thank you! I can try to do so. 
BTW, this issue occurs with the AVX2 implementation as well, when run on a Cascade Lake machine.
I don't have access to a Cascade Lake machine, so after finding the errant code, I can disassemble `libtorch_cpu.so` present in the build artifact of a CascadeLake CI machine & compare the difference with disassembled `libtorch_cpu.so` built with the same gcc version on a machine on which this test passes.

**Comment by imaginary-person** [comment]
@jerryzh168 @supriyar,

FBGEMM leverages the `AVX512_VNNI` instruction set, if supported on a platform, and Cascade Lake supports it.
Although currently ATen isn't compiled with AVX512 support, FBGEMM is.
So, the output of `test_lstm` that fails on Cascade Lake machines is a bit different from machines not supporting `AVX512_VNNI`. Fortunately, it's not a compiler bug, so I'm closing this issue. Thanks!

![CASCADE LAKE DIFFERENCE](https://user-images.githubusercontent.com/76181208/122681384-aad25c80-d1b9-11eb-834f-4385c212e4e8.PNG)

**Comment by imaginary-person** [comment]
From the discussion in https://github.com/pytorch/FBGEMM/issues/125, it seems that `reduce_range` should not be set to `True` if FBGEMM is compiled with `AVX512_VNNI` support, so reopening the issue, as it can be fixed, and this test need not be skipped.

**Comment by andrewor14** [comment]
@imaginary-person Looks like it's been 2 years. Is this still an issue with the latest master? If not can we close this?
